app/src/main/java/edu/wisc/ece/pinpoint/utils/FirebaseDriver.java

```diff
@@ -120,6 +121,28 @@ public Task<Pin> fetchPin(@NonNull String pid) {
         });
     }
 
+    public Pin getCachedPin(@NonNull String pid) {
+        return pins.get(pid);
+    }
+
+    public Task<String> dropPin(@NonNull String content, @NonNull Pin.PinType type,
```

Comment:
Small organizational suggestion, but maybe we could have `dropPin` just take in a Pin object, and within the `Pin` class we keep a `serialize` method that outputs the `Pin` in document form for firestore?

Then this `dropPin` method could just be 
```java
dropPin(Pin p) {
    return functions.getHttpsCallable("dropPin").call(p.serialize()) // etc, etc
}
```

Just feels like good practice to keep that serialization within the `Pin` class itself.

Reply:
That'd be a good idea but since we don't need to pass all of the fields to the cloud function (timestamp is computed on server and UID is in auth header) we'd be sending extra unnecessary info. This would increase our bandwidth and cost more as we scale. It's also nice to encapsulate the pin creation in FirebaseDriver to ensure it gets added to the local cache.

Reply:
The passed `Pin p` could still be appended to the local cache, e.g.

```java
return functions.getHttpsCallable("dropPin").call(p.serialize()).continueWith(task -> {
     String pid = (String) task.getResult().getData();
     pins.put(pid, p);
     return pid;
});
```

 just the serialization would happen in the `Pin` class, e.g.

```java
public class Pin {
    public Map<String, Object> serialize() {
        Map<String, Object> data = new HashMap<>();
        data.put("content", content);
        data.put("type", type.toString());
        data.put("latitude", location.getLatitude());
        data.put("longitude", location.getLongitude());
        data.put("caption", caption);
        return data;
    }
}
```

This way, if we need to add more fields to Pins, we only need to edit the `Pin` class, instead of remembering to edit `FirebaseDriver` as well.

Reply:
Implemented 💯